File: linked_list.py
class Student:
    def __init__(self, roll, name, marks):
        self.roll = roll
        self.name = name
        self.marks = marks
        self.next = None

class StudentList:
    def __init__(self):
        self.head = None

    def add(self, roll, name, marks):
        new_node = Student(roll, name, marks)
        if not self.head:
            self.head = new_node
        else:
            curr = self.head
            while curr.next:
                curr = curr.next
            curr.next = new_node

    def remove(self, roll):
        if not self.head:
            return
        if self.head.roll == roll:
            self.head = self.head.next
            return
        curr = self.head
        while curr.next and curr.next.roll != roll:
            curr = curr.next
        if curr.next:
            curr.next = curr.next.next

    def search(self, roll):
        curr = self.head
        while curr:
            if curr.roll == roll:
                print(f"{curr.roll} {curr.name} {curr.marks}")
                return
            curr = curr.next
        print("Not found")

    def display(self):
        curr = self.head
        if not curr:
            print("List is empty.")
        while curr:
            print(f"{curr.roll} {curr.name} {curr.marks}")
            curr = curr.next

    def sort_by_roll(self):
        i = self.head
        while i:
            j = i.next
            while j:
                if i.roll > j.roll:
                    i.roll, j.roll = j.roll, i.roll
                    i.name, j.name = j.name, i.name
                    i.marks, j.marks = j.marks, i.marks
                j = j.next
            i = i.next

def main():
    student_list = StudentList()

    while True:
        print("\n1. Add 2. Delete 3. Search 4. Display 5. Sort by Roll 6. Exit:")
        choice = int(input("Enter your choice: "))
        if choice == 1:
            roll = int(input("Enter roll: "))
            name = input("Enter name: ")
            marks = float(input("Enter marks: "))
            student_list.add(roll, name, marks)
        elif choice == 2:
            roll = int(input("Enter roll to delete: "))
            student_list.remove(roll)
        elif choice == 3:
            roll = int(input("Enter roll to search: "))
            student_list.search(roll)
        elif choice == 4:
            student_list.display()
        elif choice == 5:
            student_list.sort_by_roll()
            print("List sorted by roll number:")
            student_list.display()
        elif choice == 6:
            print("Exiting the program.")
            break
        else:
            print("Invalid choice")

if __name__ == "__main__":
    main()

File: linked_list.cpp
#include <bits/stdc++.h>
using namespace std;

struct Student {
    int roll;
    string name;
    float marks;
    Student* next;
};

struct StudentList {
    Student* head = nullptr;
    void add(int roll, string name, float marks) {
        Student* newNode = new Student{roll, name, marks, nullptr};
        if (!head) {
            head = newNode;
        } else {
            Student* curr = head;
            while (curr->next) {
                curr = curr->next; }
            curr->next = newNode;
 }
    }

    void remove(int roll) {
        if (!head) return;

        if (head->roll == roll) {
            Student* temp = head;
            head = head->next;
            delete temp;
            return;}
        Student* curr = head;
        while (curr->next && curr->next->roll != roll) {
            curr = curr->next;
        }
        if (curr->next) {
            Student* temp = curr->next;
            curr->next = temp->next;
            delete temp;
        }
    }

    void search(int roll) {
        Student* curr = head;
        while (curr) {
            if (curr->roll == roll) {
                cout << curr->roll << " " << curr->name << " " << curr->marks << "\n";
                return; }
            curr = curr->next;
        }
        cout << "Not found\n";
    }

    void display() {
        Student* curr = head;
        while (curr){
            cout << curr->roll << " " << curr->name << " " << curr->marks << "\n";
            curr = curr->next;}
    }

    void sortByRoll() {
        if (!head || !head->next) 
           return;
        bool swapped;
        do {swapped = false;
            Student** currPtr = &head;
            while ((*currPtr)->next) {
                Student* a = *currPtr; Student* b = a->next;
                if (a->roll > b->roll) {
                    a->next = b->next;
                    b->next = a;
                    *currPtr = b; 
                    swapped = true;}
                currPtr = &((*currPtr)->next);}
        } while (swapped);
    }

    void concatenate(StudentList& other) {
        if (!head) {
            head = other.head;
        } else {
            Student* curr = head;
            while (curr->next) {
                curr = curr->next;}
            curr->next = other.head;}
        other.head = nullptr; }};

int main() {
    StudentList list1, list2;
    int choice, roll;
    string name;
    float marks;

    do {
        cout << "1. Add to List 1\n";
        cout << "2. Delete from List 1\n";
        cout << "3. Search in List 1\n";
        cout << "4. Display List 1\n";
        cout << "5. Sort List 1 by Roll\n";
        cout << "6. Add to List 2\n";
        cout << "7. Display List 2\n";
        cout << "8. Concatenate List 2 to List 1\n";
        cout << "9. Exit\n";
        cout << "Enter choice: ";
        cin >> choice;

        if (choice == 1) {
            cout << "Enter Roll, Name, Marks: ";
            cin >> roll >> name >> marks;
            list1.add(roll, name, marks);
        } else if (choice == 2) {
            cout << "Enter Roll to delete: ";
            cin >> roll;
            list1.remove(roll);
        } else if (choice == 3) {
            cout << "Enter Roll to search: ";
            cin >> roll;
            list1.search(roll);
        } else if (choice == 4) {
            cout << "List 1:\n";
            list1.display();
        } else if (choice == 5) {
            list1.sortByRoll();
            cout << "List 1 sorted by roll number.\n";
            list1.display(); 
        } else if (choice == 6) {
            cout << "Enter Roll, Name, Marks: ";
            cin >> roll >> name >> marks;
            list2.add(roll, name, marks);
        } else if (choice == 7) {
            cout << "List 2:\n";
            list2.display();
        } else if (choice == 8) {
            list1.concatenate(list2);
            cout << "List 2 has been concatenated into List 1.\n";
        } else {
            cout << "Invalid choice. Try again.\n";
        }
    } while (choice != 9);

    return 0;
}
